File: Leetcode/Middle_Java/3sum-closest.java

// 时间复杂度 O(n^2) 空间复杂度 O(1)
// 与 3sum 一样的夹逼法，注意如果为当前的 target，直接返回
class Solution {
    public int threeSumClosest(int[] nums, int target) {
        Arrays.sort(nums);
        int ret = nums[0] + nums[1] + nums[2];
        for (int begin = 0; begin < nums.length -2; begin++){
            if (begin > 0 && nums[begin] == nums[begin-1]){
                continue;
            }
            int a = nums[begin];
            int middle = begin + 1;
            int end = nums.length - 1;
            while (middle < end){
                int b = nums[middle];
                int c= nums[end];
                int ans = target - a - b - c;
                if (Math.abs(ans) < Math.abs(ret - target)){
                    ret = a + b + c;
                }
                if (ans == 0){
                    return target;   
                }
                else if (ans > 0){
                    middle ++;
                }
                else
                    end --;    
            }
        }
        return ret;
    }
}

File: Leetcode/Middle_Python/3sum-closest.py


class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        nums.sort()
        ret = sum(nums[:3])
        for begin in range(0,len(nums)-2):
            a = nums[begin]
            if begin > 0 and nums[begin] == nums[begin-1]:
                continue
            middle = begin + 1
            end = len(nums) -1
            while middle < end:
                b = nums[middle]
                c = nums[end]
                ans = target - a - b - c
                if abs(ans) < abs(target - ret):
                    ret = a + b + c
                if ans > 0:
                    middle += 1
                elif ans == 0:
                    return target
                else:
                    end -= 1
        return ret
                
        